src/tool/hpcrun/control-knob.c

```diff
@@ -1,101 +1,135 @@
+#include <hpcrun/memory/hpcrun-malloc.h>
+
 #include "control-knob.h"
 #include <utilities/tokenize.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 
 
-static char *control_knob_names[] = {
-#define DEFINE_KNOB_NAMES(knob_name) \
-  #knob_name,
-
-  FORALL_KNOBS(DEFINE_KNOB_NAMES) 
+typedef struct control_knob {
+	struct control_knob *next;
+	char *name;
+	char *value;
+	control_knob_type type;
+} control_knob_t;
 
-#undef DEFINE_KNOB_NAMES
-};
 
+static control_knob_t *control_knobs = NULL;
 
-typedef struct control_knob {
-  char *name;
-  char *value;
-} control_knob_t;
 
-control_knob_t control_knobs[HPCRUN_NUM_CONTROL_KNOBS];
+static control_knob_t *
+control_knob_name_lookup(char *in)
+{
+	control_knob_t *iter = control_knobs;
 
+	while (iter != NULL){
+		if(strcmp(iter->name, in) == 0) return iter;
 
-static void control_knob_process_string(char *in);
+		iter = iter->next;
+	}
+	return NULL;
+}
 
 
 void
-control_knob_init()
+control_knob_register(char *name, char *value, control_knob_type type)
 {
-  char *s = getenv("HPCRUN_CONTROL_KNOBS");
-  if (s) {
-#define INIT_KNOBS(knob_name) \
-  control_knobs[knob_name].name = control_knob_names[knob_name]; \
-  control_knobs[knob_name].value = NULL;
-
-  FORALL_KNOBS(INIT_KNOBS) 
+	control_knob_t *iter = control_knob_name_lookup(name);
+
+	if (iter == NULL){
+		iter = (control_knob_t*) malloc(sizeof(control_knob_t));
+		iter->name = strdup(name);
+		iter->type = type;
+	}
+	iter->value = strdup(value);
+	iter->next = control_knobs;
+	control_knobs = iter;
+}
 
-#undef INIT_KNOBS
 
-    control_knob_process_string(s);
-  }
+static void
+control_knob_default_register(){
+	control_knob_register("STREAMS_PER_THREAD", "65536", ck_int);
+	control_knob_register("MAX_THREADS_CONSUMERS", "256", ck_int);
 }
 
 
-char *
-control_knob_value_get(control_category c)
+void
+control_knob_init()
 {
-  if (c < HPCRUN_NUM_CONTROL_KNOBS) {
-    return control_knobs[c].value;
-  }
-  return NULL;
+	control_knob_default_register();
```

Comment:
It would be better if you can change the indentation space from four to two.

Reply:
I would, but in my editor it is already on two. I am not sure why it makes it 4 on git. I could probably change indentation from llnl.